tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AdoNet/DbScopeFactory.cs

```diff
@@ -128,19 +140,45 @@ public static Scope CreateDbCommandScope(Tracer tracer, IDbCommand command)
                 if (commandType == CommandType)
                 {
                     // use the cached values if command.GetType() == typeof(TCommand)
-                    return DbScopeFactory.CreateDbCommandScope(tracer, command, IntegrationId, DbTypeName, OperationName);
+                    return DbScopeFactory.CreateDbCommandScope(tracer, command, IntegrationId, DbTypeName, OperationName, GetServiceName(tracer, DbTypeName));
                 }
 
                 // if command.GetType() != typeof(TCommand), we are probably instrumenting a method
                 // defined in a base class like DbCommand and we can't use the cached values
                 if (TryGetIntegrationDetails(commandType.FullName, out var integrationId, out var dbTypeName))
                 {
                     var operationName = $"{dbTypeName}.query";
-                    return DbScopeFactory.CreateDbCommandScope(tracer, command, integrationId.Value, dbTypeName, operationName);
+                    return DbScopeFactory.CreateDbCommandScope(tracer, command, integrationId.Value, dbTypeName, operationName, GetServiceName(tracer, dbTypeName));
                 }
 
                 return null;
             }
+
+            private static string GetServiceName(Tracer tracer, string dbTypeName)
+            {
+                if (!tracer.Settings.TryGetServiceName(dbTypeName, out string serviceName))
+                {
+                    if (_tracerDefaultServiceName == tracer.DefaultServiceName)
+                    {
+                        serviceName = _serviceName;
+                    }
+                    else
+                    {
+                        if (_tracerDefaultServiceName is null)
+                        {
+                            _tracerDefaultServiceName = tracer.DefaultServiceName;
+                            serviceName = $"{_tracerDefaultServiceName}-{dbTypeName}";
+                            _serviceName = serviceName;
+                        }
+                        else
+                        {
+                            serviceName = $"{_tracerDefaultServiceName}-{dbTypeName}";
+                        }
+                    }
+                }
+
+                return serviceName;
+            }
```

Comment:
I think we can improve (and simplify) this slightly? It looks like we're accounting for the fact that the default service name can change (due to tracer settings changing)? But in that case I think we can overwrite `_tracerDefaultServiceName` with the "new" default, as we always completely replace the settings. i.e. all new requests will have the new default service name, we won't flip flop back and forth?

```suggestion
            private static string GetServiceName(Tracer tracer, string dbTypeName)
            {
                if (!tracer.Settings.TryGetServiceName(dbTypeName, out string serviceName))
                {
                    if (_tracerDefaultServiceName == tracer.DefaultServiceName)
                    {
                        serviceName = _serviceName;
                    }
                    else
                    {
                        _tracerDefaultServiceName = tracer.DefaultServiceName;
                        serviceName = $"{_tracerDefaultServiceName}-{dbTypeName}";
                        _serviceName = serviceName;
                    }
                }

                return serviceName;
            }
```

Reply:
I've changed the algorithm in 769ed59